### PR TITLE
UCT/TCP/EP: Implement AM short send

### PR DESCRIPTION
**Pull request by dmitrygladkov**
## What

This PR implements AM short send

## Why ?

This improves unidirectional message rate by ~5-7%

## How ?

Implemented common `uct_tcp_ep_am_send` function that performs send of internal TCP AM header and AM data.

### CODE REVIEW

**Comment by dmitrygladkov** [comment]
bot:bgate:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8791/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8799/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8803/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
@hoopoepg @brminich @evgeny-leksikov can someone take a look plz?

File: src/uct/tcp/tcp_ep.c [hunk 1]
```diff
@@ -257,37 +257,73 @@ unsigned uct_tcp_ep_progress_rx(uct_tcp_ep_t *ep)
...
 
+static inline ucs_status_t uct_tcp_ep_am_send(uct_tcp_iface_t *iface, uct_tcp_ep_t *ep,
+                                              const uct_tcp_am_hdr_t *hdr,
+                                              size_t check_length, const char *name)
+{
+    UCT_CHECK_LENGTH(hdr->length, 0, check_length, name);
+    uct_iface_trace_am(&iface->super, UCT_AM_TRACE_TYPE_SEND, hdr->am_id,
+                       hdr + 1, hdr->length, "SEND fd %d", ep->fd);
+
+    ep->length  = sizeof(*hdr) + hdr->length;
```
Comment (hoopoepg) [inline src/uct/tcp/tcp_ep.c]: align '=' with line below

File: src/uct/tcp/tcp_ep.c [hunk 2]
```diff
@@ -257,37 +257,73 @@ unsigned uct_tcp_ep_progress_rx(uct_tcp_ep_t *ep)
...
+    if (ep->length > 0) {
+        uct_tcp_ep_mod_events(ep, EPOLLOUT, 0);
+    }
+    return UCS_OK;
+}
+
+ucs_status_t uct_tcp_ep_am_short(uct_ep_h uct_ep, uint8_t am_id, uint64_t header,
+                                 const void *payload, unsigned length)
+{
+    uct_tcp_ep_t *ep = ucs_derived_of(uct_ep, uct_tcp_ep_t);
```
Comment (hoopoepg) [inline src/uct/tcp/tcp_ep.c]: align variable name & '='

File: src/uct/tcp/tcp_ep.c [hunk 3]
```diff
@@ -257,37 +257,73 @@ unsigned uct_tcp_ep_progress_rx(uct_tcp_ep_t *ep)
     return recv_length > 0;
 }
 
+static inline ucs_status_t uct_tcp_ep_am_send(uct_tcp_iface_t *iface, uct_tcp_ep_t *ep,
+                                              const uct_tcp_am_hdr_t *hdr,
+                                              size_t check_length, const char *name)
+{
+    UCT_CHECK_LENGTH(hdr->length, 0, check_length, name);
```
Comment (brminich) [inline src/uct/tcp/tcp_ep.c]: better check it before all other checks, i.e in the very beginning of in _short, _bcopy funcs

File: src/uct/tcp/tcp_ep.c [hunk 4]
```diff
@@ -257,37 +257,73 @@ unsigned uct_tcp_ep_progress_rx(uct_tcp_ep_t *ep)
...
+    uct_tcp_iface_t *iface = ucs_derived_of(uct_ep->iface, uct_tcp_iface_t);
+    uct_tcp_am_hdr_t *hdr;
+
+    UCT_CHECK_AM_ID(am_id);
+
+    if (!uct_tcp_ep_can_send(ep)) {
+        return UCS_ERR_NO_RESOURCE;
+    }
+
+    hdr         = ep->buf;
```
Comment (brminich) [inline src/uct/tcp/tcp_ep.c]: these two lines could be a part of `uct_tcp_ep_am_send` as well

File: src/uct/tcp/tcp_ep.c [hunk 5]
```diff
@@ -257,37 +257,73 @@ unsigned uct_tcp_ep_progress_rx(uct_tcp_ep_t *ep)
...
+    hdr->am_id  = am_id;
+    hdr->length = sizeof(header) + length;
+
+    memcpy(hdr + 1, &header, sizeof(header));
+    memcpy((uint8_t*)(hdr + 1) + sizeof(header), payload, length);
+
+    UCT_TL_EP_STAT_OP(&ep->super, AM, SHORT, hdr->length);
```
Comment (brminich) [inline src/uct/tcp/tcp_ep.c]: `uct_tcp_ep_am_send` below can fail, need to add stats in case of success only

File: src/uct/tcp/tcp_ep.c [hunk 6]
```diff
@@ -257,37 +257,73 @@ unsigned uct_tcp_ep_progress_rx(uct_tcp_ep_t *ep)
...
-    uct_iface_trace_am(&iface->super, UCT_AM_TRACE_TYPE_SEND, hdr->am_id,
-                       hdr + 1, hdr->length, "SEND fd %d", ep->fd);
-    iface->outstanding += ep->length;
 
-    uct_tcp_ep_send(ep);
-    if (ep->length > 0) {
-        uct_tcp_ep_mod_events(ep, EPOLLOUT, 0);
-    }
-    return packed_length;
+    uct_tcp_ep_am_send(iface, ep, hdr,
```
Comment (brminich) [inline src/uct/tcp/tcp_ep.c]: `uct_tcp_ep_am_send` can fail - need to check status. (also stats needs to be increased in case of success only)

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8808/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
> A couple of general questions:
> 
> * Is msg rate better because of eliminating 1 func call?

Discussed with @brminich internally: more simple UCP fast path for `uct_ep_am_short` contributes to this as well.

> * Is latency affected somehow?

I see no any improvements or no any degradations with this change comparing with master.

> * Maybe it's worth to reconsider default value of MAX_SHORT config for TCP?

Yes, it should be done. increasing `UCX_AM_SHORT` gives better numbers (sometimes msg rates improvement is up to 24% - 1 Kbytes):

bytes | ratio (higher is better)
-- | --
0 | 1.077139
1 | 1.090136
2 | 1.07147
4 | 1.130736
8 | 1.086909
16 | 1.073117
32 | 1.029759
64 | 1.155889
128 | 1.04798
256 | 1.154537
512 | 1.007544
1024 | 1.248267
2048 | 1.087244
4096 | 1.021886

I'll do more testing and adjust to the appropriate value - it looks that MAX_SHORT=MAX_BCOPY is the best one

@brminich @hoopoepg the PR is ready for the next round of the review. plz, review

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8815/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8818/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
@yosefe could you take a look, plz?

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 7]
```diff
@@ -295,6 +295,7 @@ ucs_status_t uct_dc_mlx5_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t hdr,
 #if HAVE_IBV_EXP_DM
     }
 
+    UCT_CHECK_AM_ID(id);
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c]: fixing ud/dc should be in another PR

File: src/uct/tcp/tcp_iface.c [hunk 8]
```diff
@@ -286,6 +289,8 @@ static UCS_CLASS_INIT_FUNC(uct_tcp_iface_t, uct_md_h md, uct_worker_h worker,
     self->outstanding            = 0;
     self->config.buf_size        = config->super.max_bcopy +
                                    sizeof(uct_tcp_am_hdr_t);
+    self->config.short_size      = config->super.max_short +
```
Comment (yosefe) [inline src/uct/tcp/tcp_iface.c]: do we want to increase the default for TCP to more than 128?

File: src/uct/tcp/tcp.h [hunk 9]
```diff
@@ -16,6 +16,41 @@
...
+        _target_length = sizeof(_am_header) + _payload_length; \
+    } while (0)
+
+#define UCT_TCP_AM_BCOPY_PACK_DATA(_pack_f, _target_buf, _target_length, \
+                                   _am_arg, ...) \
+    do { \
+        _target_length = _pack_f(_target_buf, _am_arg);	\
+    } while (0)
+
+#define UCT_TCP_AM_PREPARE(_iface, _ep, _id, _len_thr, _hdr, \
```
Comment (yosefe) [inline src/uct/tcp/tcp.h]: where is _iface used?

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 7 above)
Comment (dmitrygladkov) [reply src/uct/ib/dc/dc_mlx5_ep.c]: #3234 has been created to fix this comment

File: src/uct/tcp/tcp_iface.c (hunk 8 above)
Comment (dmitrygladkov) [reply src/uct/tcp/tcp_iface.c]: I've done more testing. We can use am_short = am_bcopy - it improves message rates, but I still don't see any performance improvements for latency

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8823/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8825/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8826/ for details (Mellanox internal link).

File: src/uct/tcp/tcp_iface.c (hunk 8 above)
Comment (yosefe) [reply src/uct/tcp/tcp_iface.c]: so let's increase it, even if it's just MR

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8831/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8833/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@dmitrygladkov can u pls squash?
also do we need to increase max_short?

**Comment by dmitrygladkov** [comment]
@yosefe 

> can u pls squash?

done

> also do we need to increase max_short?

discussed internally (done in https://github.com/openucx/ucx/commit/b4b85481820ced7999259b7df1da85eaaf60636d)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8860/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
@yosefe the testing is done due to problem unrelated to TCP

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8867/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8880/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
@brminich @hoopoepg

File: src/uct/tcp/tcp.h [hunk 10]
```diff
@@ -16,6 +16,41 @@
...
+    do { \
+        _pack_f(_target_buf, &(_am_header), sizeof(_am_header)); \
+        _pack_f((uint8_t*)(_target_buf) + sizeof(_am_header), \
+                _am_payload, _payload_length); \
+        _target_length = sizeof(_am_header) + _payload_length; \
+    } while (0)
+
+#define UCT_TCP_AM_BCOPY_PACK_DATA(_pack_f, _target_buf, _target_length, \
+                                   _am_arg, ...) \
+    do { \
```
Comment (hoopoepg) [inline src/uct/tcp/tcp.h]: minor: do you need do/while around single operation?

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 11]
```diff
@@ -295,7 +295,6 @@ ucs_status_t uct_dc_mlx5_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t hdr,
 #if HAVE_IBV_EXP_DM
     }
 
-    UCT_CHECK_AM_ID(id);
```
Comment (hoopoepg) [inline src/uct/ib/dc/dc_mlx5_ep.c]: restore please :)

File: src/uct/tcp/tcp.h (hunk 10 above)
Comment (dmitrygladkov) [reply src/uct/tcp/tcp.h]: 👍 there is no need -- fixed

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8883/ for details (Mellanox internal link).

File: src/uct/tcp/tcp_iface.c [hunk 12]
```diff
@@ -276,20 +279,23 @@ static UCS_CLASS_INIT_FUNC(uct_tcp_iface_t, uct_md_h md, uct_worker_h worker,
...
-    self->config.buf_size        = config->super.max_bcopy +
-                                   sizeof(uct_tcp_am_hdr_t);
-    self->config.prefer_default  = config->prefer_default;
-    self->config.max_poll        = config->max_poll;
-    self->sockopt.nodelay        = config->sockopt_nodelay;
-    self->sockopt.sndbuf         = config->sockopt_sndbuf;
+    self->outstanding           = 0;
+    self->config.buf_size       = config->super.max_bcopy +
+                                  sizeof(uct_tcp_am_hdr_t);
+    self->config.short_size     = ucs_max(config->super.max_short,
```
Comment (brminich) [inline src/uct/tcp/tcp_iface.c]: does this mean that MAX_SHORT var will not work properly for TCP? Maybe it's better to set bigger MAX_SHORT by default for TCP? (can be even equal to max_bcopy).

File: src/uct/tcp/tcp.h [hunk 13]
```diff
@@ -16,6 +16,39 @@
 /** How many events to wait for in epoll_wait */
 #define UCT_TCP_MAX_EVENTS        32
 
+#define UCT_TCP_AM_SHORT_PACK_DATA(_pack_f, _target_buf, _target_length, \
+                                   _am_payload, _payload_length, _am_header) \
+    do { \
+        _pack_f(_target_buf, &(_am_header), sizeof(_am_header)); \
```
Comment (brminich) [inline src/uct/tcp/tcp.h]: could just assign header in `uct_tcp_ep_am_short` (instead of extra memcpy) and have one general macro instead of 3.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8884/ for details (Mellanox internal link).

File: src/uct/tcp/tcp.h (hunk 13 above)
Comment (dmitrygladkov) [reply src/uct/tcp/tcp.h]: done

File: src/uct/tcp/tcp_iface.c (hunk 12 above)
Comment (dmitrygladkov) [reply src/uct/tcp/tcp_iface.c]: done by overriding `UCX_TCP_MAX_SHORT` env

**Comment by dmitrygladkov** [comment]
bot:bgate:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8956/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
test:all:retest

**Comment by dmitrygladkov** [comment]
test:mlx:retest

**Comment by dmitrygladkov** [comment]
test:bgate:retest

**Comment by dmitrygladkov** [comment]
bot:retest

**Comment by dmitrygladkov** [comment]
@brminich I fixed your comments - could you take a look plz?

File: src/uct/tcp/tcp_ep.c [hunk 14]
```diff
@@ -8,6 +8,41 @@
...
+        \
+        (_hdr)        = (_ep)->buf; \
+        (_hdr)->am_id = _id; \
+        \
+        UCT_TCP_AM_ ## _method ## _PACK_DATA(_pack_f, (_hdr) + 1, (_hdr)->length, \
+                                             _am_payload, _payload_length, \
+                                             _am_header); \
+        \
+        UCT_CHECK_LENGTH((_hdr)->length, 0, _len_thr, _name); \
+        UCT_TL_EP_STAT_OP(&(_ep)->super, AM, _method, (_hdr)->length); \
```
Comment (brminich) [inline src/uct/tcp/tcp_ep.c]: stats needs to be increased when send is sucesfull, 
`uct_tcp_ep_am_send` (invoked after) can fail

File: src/uct/tcp/tcp_ep.c [hunk 15]
```diff
@@ -8,6 +8,41 @@
...
+        UCT_CHECK_AM_ID(_id); \
+        \
+        if (!uct_tcp_ep_can_send(_ep)) { \
+            return UCS_ERR_NO_RESOURCE; \
+        } \
+        \
+        (_hdr)        = (_ep)->buf; \
+        (_hdr)->am_id = _id; \
+        \
+        UCT_TCP_AM_ ## _method ## _PACK_DATA(_pack_f, (_hdr) + 1, (_hdr)->length, \
```
Comment (brminich) [inline src/uct/tcp/tcp_ep.c]: imo, macro looks too generic, while it is not supposed to be used in other transports.
I'd go with short common inline function checking resources and filling the header, and the rest would do in the TL functions itself

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8968/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8965/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8969/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8975/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
infra issue
should I retest mlx? @yosefe

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8978/ for details (Mellanox internal link).
